vllm/distributed/kv_transfer/kv_lookup_buffer/simple_dict_buffer.py

```diff
@@ -0,0 +1,136 @@
+"""
+    Implements a distributed key-value (KV) cache transfer mechanism.
+
+    Key Features:
+    - Distributed KV cache transmission using PyNccl pipes.
+    - Non-blocking `insert`, blocking `drop_select`.
+    - Use CPU signal pipe to avoid racing condition
+    - Handles buffer size constraints and provide backpressure mechanism to 
+      stop the prefill instance when the decode instance is slow.
+"""
+import threading
+from typing import Dict, List, Optional
+
+import torch
+
+from vllm.distributed.kv_transfer.kv_lookup_buffer.base import (
+    KVLookupBufferBase)
+from vllm.distributed.kv_transfer.kv_pipe.base import KVPipeBase
+from vllm.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+def _string_to_byte_tensor(s: str) -> torch.Tensor:
+    byte_data = s.encode('utf-8')
+    byte_list = list(byte_data)
+    byte_tensor = torch.tensor(byte_list, dtype=torch.uint8, device='cpu')
+    return byte_tensor
+
+
+def _byte_tensor_to_string(t: torch.Tensor) -> str:
+    byte_list = t.tolist()
+    byte_data = bytes(byte_list)
+    return byte_data.decode('utf-8')
+
+
+class SimpleDictBuffer(KVLookupBufferBase):
+
+    def __init__(self, signal_pipe: KVPipeBase, data_pipe: KVPipeBase,
+                 buffer_size_thresh: float):
+        """
+        signal_pipe: on CPU 
+        
+        NOTE: on-device recv will block all threads in the process, making the 
+        KV cache producer unable to listen to new request while transmitting 
+        KV cache. Luckily CPU recv only blocks the current thread so we use 
+        CPU recv to listen to new request.
+        
+        data_pipe: on device (e.g. GPU)
+        """
+
+        self.buffer: Dict[str, List[torch.Tensor]] = {}
+
+        self.buffer_size = 0
+        self.buffer_size_threshold = buffer_size_thresh
+        self.buffer_lock = threading.Lock()
+        self.signal_pipe = signal_pipe
+        self.data_pipe = data_pipe
+        self.request_handling_thread: Optional[threading.Thread] = None
+
+        self.normal_signal = torch.tensor([0], device="cpu")
+        self.end_signal = None
+
+    def _is_end_signal(self, signal):
+        return signal is None
+
+    def drop_select_handler(self):
+
+        try:
+
+            while True:
+                signal = self.signal_pipe.recv_tensor()
+                if self._is_end_signal(signal):
+                    logger.info("Received end signal!")
+                    break
+
+                key_bytes = self.data_pipe.recv_tensor()
+                key_hash = _byte_tensor_to_string(key_bytes)
+
+                with self.buffer_lock:
```

Comment:
I think self.buffer_block covers too much
 it should only cover self.buffer.  sometimes  pynccl_pipe.send_tensor will blocking when self.block_if_full() is called,

Reply:
You are right, unnecessary to lock the case of send_tensor(None). Will update.